core/src/client/mod.rs

```diff
@@ -468,14 +475,113 @@ pub enum IdKind {
 }
 
 impl IdKind {
-	fn into_id(self, id: u64) -> Id<'static> {
+	/// Generate an `Id` from number.
+	pub fn into_id(self, id: u64) -> Id<'static> {
 		match self {
 			IdKind::Number => Id::Number(id),
 			IdKind::String => Id::Str(format!("{}", id).into()),
 		}
 	}
 }
 
+/// Generate a range of IDs to be used in a batch request.
+pub fn generate_batch_id_range(guard: &RequestIdGuard<Id>, len: u64) -> Result<Range<u64>, Error> {
+	let id_start = guard.inner().try_parse_inner_as_number().unwrap();
+	let id_end = id_start
+		.checked_add(len)
+		.ok_or_else(|| Error::Custom("BatchID range is wrapped; restart client or try again later".to_string()))?;
+
+	Ok(id_start..id_end)
+}
+
+/// Represent a single entry in a batch response.
+pub type BatchEntry<'a, R> = Result<R, ErrorObject<'a>>;
+
+/// Batch response.
+#[derive(Debug)]
+pub struct BatchResponse<'a, R> {
+	successful_calls: usize,
+	failed_calls: usize,
+	responses: Vec<BatchEntry<'a, R>>,
+}
+
+impl<'a, R: fmt::Debug + 'a> BatchResponse<'a, R> {
+	/// Create a new [`BatchResponse`].
+	pub fn new(successful_calls: usize, responses: Vec<BatchEntry<'a, R>>, failed_calls: usize) -> Self {
+		Self { successful_calls, responses, failed_calls }
+	}
+
+	/// Get the length of the batch response.
+	pub fn len(&self) -> usize {
+		self.responses.len()
+	}
+
+	/// Is empty.
+	pub fn is_empty(&self) -> bool {
+		self.responses.len() == 0
+	}
+
+	/// Get the number of successful calls in the batch.
+	pub fn num_successful_calls(&self) -> usize {
+		self.successful_calls
+	}
+
+	/// Get the number of failed calls in the batch.
+	pub fn num_failed_calls(&self) -> usize {
+		self.failed_calls
+	}
+
+	/// Returns `Ok(iterator)` if all responses were successful
+	/// otherwise `Err(iterator)` is returned.
+	///
+	/// If you want get all responses if an error responses occurs use [`BatchResponse::into_iter`]
+	/// instead where it's possible to implement customized logic.
+	pub fn ok(
```

Comment:
I wonder whether the names should be `ok` and `into_ok` to mirror `iter` and `into_iter`?